### PR TITLE
Parallel Compression

### PR DESCRIPTION
**Pull request by jwong-nd**
Logic:
- Decoupled logic for recursive copy from convert_video calls for parallelization
- Filesystem.py logic moved into etl.py to reference fields in self.job_settings directly
- Lift logging outside of convert_video for parallelization

Readibility:
- Added comments, added typing, changed some variable names

### CODE REVIEW

File: src/aind_behavior_video_transformation/etl.py [hunk 1]
```diff
@@ -64,6 +77,137 @@ class BehaviorVideoJob(GenericEtl[BehaviorVideoJobSettings]):
...
+            # Add to job buffer
+            convert_video_params.append((vid_path, output_vid_dir, arg_set))
+            logging.info(
+                f"Compressing {str(vid_path)} \
+                         w/ {comp_req.compression_enum}"
+            )
+
+        if self.job_settings.parallel_compression:
+            # Dask implementation
+            # import dask
```
Comment (jtyoung84) [inline src/aind_behavior_video_transformation/etl.py:187]: We can remove the commented out code if it's not being used

File: pyproject.toml [hunk 2]
```diff
@@ -59,7 +59,7 @@ exclude_lines = [
   "import",
   "pragma: no cover",
 ]
-fail_under = 97
+fail_under = 80
```
Comment (jtyoung84) [inline pyproject.toml:62]: We shouldn't be lowering the coverage percentage. Some of the methods added like `_resolve_compression_requests` and `_format_output_directory` can easily have corresponding tests.

File: src/aind_behavior_video_transformation/transform_videos.py [hunk 3]
```diff
@@ -204,17 +222,16 @@ def convert_video(video_path: Path, dst: Path, arg_set) -> Path:
     output_args = arg_set[1]
 
     ffmpeg_command = ["ffmpeg", "-y", "-v", "warning", "-hide_banner"]
+    # Use 8 threads per compression job
+    ffmpeg_command.extend(["-threads", "8"])
```
Comment (jtyoung84) [inline src/aind_behavior_video_transformation/transform_videos.py:226]: It might be helpful to make this configurable with 8 set as the default.

Comment (galenlynch) [reply src/aind_behavior_video_transformation/transform_videos.py:226]: We should not set this by default... ffmpeg will use all cores by default. Specifying a number less than zero reduces the number of cores each process can use.

File: tests/test_transform_videos.py [hunk 4]
```diff
@@ -113,7 +110,9 @@ def test_run_job_with_data_structure(self, mock_time: MagicMock):
             camera_in_paths = [in_path / d for d in camera_subdirs]
             for camera_path in camera_in_paths:
                 camera_path.mkdir()
-                symlink(test_vid_path, camera_path / test_vid_name)
+                shutil.copy(
```
Comment (galenlynch) [inline tests/test_transform_videos.py:113]: Why copy instead of symlink?

File: src/aind_behavior_video_transformation/etl.py [hunk 5]
```diff
@@ -43,6 +43,19 @@ class BehaviorVideoJobSettings(BasicJobSettings):
             "request"
         ),
     )
+    parallel_compression: bool = Field(
+        default=True,
+        description="Run compression in parallel or sequentially.",
+    )
+    video_extensions: List[str] = [
```
Comment (galenlynch) [inline src/aind_behavior_video_transformation/etl.py:50]: This could be a constant someplace, but I guess it's ok as a job parameter

File: src/aind_behavior_video_transformation/etl.py [hunk 6]
```diff
@@ -64,6 +77,137 @@ class BehaviorVideoJob(GenericEtl[BehaviorVideoJobSettings]):
...
+                            override_file.suffix.lower()
+                            in self.job_settings.video_extensions
+                        ):
+                            override_file = override_file.resolve()
+                            overrides[override_file] = override_req
+
+            # Filter path_req_pairs by overrides.
+            output_pairs: List[Tuple[Path, CompressionRequest]] = []
+            for vid_path, default_req in path_req_pairs:
+                for override_path, override_req in overrides.items():
```
Comment (galenlynch) [inline src/aind_behavior_video_transformation/etl.py:149]: You can just check if `vid_path` is in `overrides`.

File: src/aind_behavior_video_transformation/etl.py [hunk 7]
```diff
@@ -64,6 +77,137 @@ class BehaviorVideoJob(GenericEtl[BehaviorVideoJobSettings]):
...
+                jobs = [
+                    executor.submit(convert_video, *params)
+                    for params in convert_video_params
+                ]
+                for job in as_completed(jobs):
+                    try:
+                        result = job.result()
+                        logging.info("FFmpeg job completed:", result)
+                    except Exception as e:
+                        logging.info("Error:", e)
```
Comment (galenlynch) [inline src/aind_behavior_video_transformation/etl.py:204]: This is swallowing errors. I strongly think any error should cause the entire job to error.

**Review by galenlynch — CHANGES_REQUESTED** [review]
Can you please split this PR into the following distinct PRs:

1) a PR that is just related to parallelization
2) a PR that moves functions into class methods
3) a PR that changes the logic of testing for overrides etc

Too many things are happening in this PR, and because you've moved functions around it's hard to tell what's changed.

File: src/aind_behavior_video_transformation/etl.py (hunk 6 above)
Comment (galenlynch) [reply src/aind_behavior_video_transformation/etl.py:149]: I guess that's not true if they're not all normalized.

**Review by galenlynch — COMMENTED** [review]
I think this is pretty close. I haven't had a chance to run the PR yet, but did you confirm that the information from each process, printed by ffmpeg to stderr, is being combined correctly? I would have thought that you would need to redirect the stderr of each subprocess if you were invoking them in parallel, so that you could log it or print it back out when the future completes.

It wasn't clear to me why you exported `convert_video` at the top level (in the __init__ file), or why you added another test of it. Each of the other tests should also test `convert_video`

File: src/aind_behavior_video_transformation/transform_videos.py [hunk 8]
```diff
@@ -204,17 +226,17 @@ def convert_video(video_path: Path, dst: Path, arg_set) -> Path:
     output_args = arg_set[1]
 
     ffmpeg_command = ["ffmpeg", "-y", "-v", "warning", "-hide_banner"]
+
+    # Set thread count
+    ffmpeg_command.extend(["-threads", str(ffmpeg_thread_cnt)])
```
[the remark this answers is not in the document]
Comment (galenlynch) [reply src/aind_behavior_video_transformation/transform_videos.py:231]: Also do you test setting this?

File: tests/test_transform_videos.py [hunk 9]
```diff
@@ -76,6 +99,7 @@ def test_run_job(self, mock_time: MagicMock):
                     compression_requested=CompressionRequest(
                         compression_enum=compression_enum
                     ),
+                    parallel_compression=False,
```
[the remark this answers is not in the document]
Comment (jwong-nd) [reply tests/test_transform_videos.py:102]: By default, parallel_compression is set to True, so parallelization is tested in the directory setting 'test_run_job_with_data_structure'

File: tests/test_transform_videos.py [hunk 10]
```diff
@@ -55,6 +53,31 @@ class TestBehaviorVideoJob(unittest.TestCase):
     test_vid_name = "clip.mp4"
     test_vid_path = test_data_path / test_vid_name
 
+    @patch("aind_behavior_video_transformation.etl.time")
+    def test_convert_video(self, mock_time: MagicMock):
```
[the remark this answers is not in the document]
Comment (jwong-nd) [reply tests/test_transform_videos.py:57]: Yes, this is tested by the other functions, though the coverage library does not recognize this (don't know why) and then the coverage report fails the 97% threshold. I imported the function and added a test for that

[the remark this answers is not in the document]
Comment (jwong-nd) [reply tests/test_transform_videos.py:57]: I still cannot run the tests locally to produce such a report so if you want to investigate this feel free...

File: src/aind_behavior_video_transformation/__init__.py [hunk 11]
```diff
@@ -9,4 +9,5 @@
 from aind_behavior_video_transformation.transform_videos import (  # noqa F401
     CompressionEnum,
     CompressionRequest,
+    convert_video,
```
[the remark this answers is not in the document]
Comment (jwong-nd) [reply src/aind_behavior_video_transformation/__init__.py:12]: See below

File: src/aind_behavior_video_transformation/transform_videos.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jwong-nd) [reply src/aind_behavior_video_transformation/transform_videos.py:231]: in next batch of commits

File: src/aind_behavior_video_transformation/transform_videos.py [hunk 12]
```diff
@@ -1,6 +1,17 @@
-"""Module to handle transforming behavior videos"""
+"""
+Module to handle transforming behavior videos
+
+To add a new compression preset:
+1) Define a CompressionEnum: 'CUSTOM_COMPRESSION = 'custom'
```
[the remark this answers is not in the document]
Comment (jwong-nd) [reply src/aind_behavior_video_transformation/transform_videos.py:5]: in next commit batch

File: tests/test_transform_videos.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jwong-nd) [reply tests/test_transform_videos.py:102]: Added explicit flag in 'test_run_job_with_data_structure'

File: src/aind_behavior_video_transformation/etl.py [hunk 13]
```diff
@@ -64,6 +73,37 @@ class BehaviorVideoJob(GenericEtl[BehaviorVideoJobSettings]):
...
+                    executor.submit(
+                        convert_video,
+                        *params,
+                        self.job_settings.ffmpeg_thread_cnt,
+                    )
+                    for params in convert_video_args
+                ]
+                for job in as_completed(jobs):
+                    result = job.result()
+                    logging.info("FFmpeg job completed:", result)
```
[the remark this answers is not in the document]
Comment (jwong-nd) [reply src/aind_behavior_video_transformation/etl.py:98]: This is the output of a working call (same input directory as test_run_job_with_data_structure):
"""
2024-11-07 11:04:40,762 - INFO - FFmpeg job completed: /private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmpxq15v823/camera2/clip.mp4
2024-11-07 11:04:40,792 - INFO - FFmpeg job completed: /private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmpxq15v823/camera1/clip.mp4
"""

This is the output if I set the requested compression to:
"""
faulty_req = CompressionRequest(
        compression_enum=CompressionEnum.USER_DEFINED,
        user_ffmpeg_input_options = "invalid user input args",
        user_ffmpeg_output_options = "invalid user output args",
)
"""

Nothing is logged and I believe stderr is forwarded to stdout:
"""
[AVFormatContext @ 0x11df07720] Unable to choose an output format for 'invalid'; use a standard extension for the filename or specify the format manually.
[AVFormatContext @ 0x12e006c50] [out#0 @ 0x11df12660] Unable to choose an output format for 'invalid'; use a standard extension for the filename or specify the format manually.
Error initializing the muxer for invalid: Invalid argument
Error opening output file invalid.
[out#0 @ 0x12e011b70] Error opening output files: Invalid argument
Error initializing the muxer for invalid: Invalid argument
Error opening output file invalid.
Error opening output files: Invalid argument
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/Users/jonathan.wong/miniconda3/envs/compression/lib/python3.11/concurrent/futures/process.py", line 261, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jonathan.wong/Projects/aind-behavior-video-transformation/src/aind_behavior_video_transformation/transform_videos.py", line 242, in convert_video
    subprocess.run(ffmpeg_command, check=True)
  File "/Users/jonathan.wong/miniconda3/envs/compression/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['ffmpeg', '-y', '-v', 'warning', '-hide_banner', 'invalid', 'user', 'input', 'args', '-i', '/private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmp7ogiojea/camera2/clip.mp4', 'invalid', 'user', 'output', 'args', '/private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmpxj4mbyd_/camera2/clip.mp4']' returned non-zero exit status 234.
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/jonathan.wong/Projects/aind-behavior-video-transformation/src/aind_behavior_video_transformation/etl.py", line 234, in <module>
    job_response = job.run_job()
                   ^^^^^^^^^^^^^
  File "/Users/jonathan.wong/Projects/aind-behavior-video-transformation/src/aind_behavior_video_transformation/etl.py", line 140, in run_job
    self._run_compression(convert_video_args)
  File "/Users/jonathan.wong/Projects/aind-behavior-video-transformation/src/aind_behavior_video_transformation/etl.py", line 97, in _run_compression
    result = job.result()
             ^^^^^^^^^^^^
  File "/Users/jonathan.wong/miniconda3/envs/compression/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/jonathan.wong/miniconda3/envs/compression/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
subprocess.CalledProcessError: Command '['ffmpeg', '-y', '-v', 'warning', '-hide_banner', 'invalid', 'user', 'input', 'args', '-i', '/private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmp7ogiojea/camera2/clip.mp4', 'invalid', 'user', 'output', 'args', '/private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmpxj4mbyd_/camera2/clip.mp4']' returned non-zero exit status 234.
"""

Is this garbled? It appears like this returns an error for any failed process.

[the remark this answers is not in the document]
Comment (galenlynch) [reply src/aind_behavior_video_transformation/etl.py:98]: It looks like the two outputs are being interleaved in a way that makes it hard to follow, for example:
```
[AVFormatContext @ 0x11df07720] Unable to choose an output format for 'invalid'; use a standard extension for the filename or specify the format manually.
[AVFormatContext @ 0x12e006c50] [out#0 @ 0x11df12660] Unable to choose an output format for 'invalid'; use a standard extension for the filename or specify the format manually.
```
IMO it would be better if the output of each process was captured and printed to stdout or stderr in a block, so you can follow the messages from each subprocess.

**Comment by jwong-nd** [comment]
Coverage threshold decreased to 93 as coverage library is not registering error handling code as run when it is. 

Extra info about logging update: 
- Stdout for 1 or 2 invalid requests: 
```
File "/Users/jonathan.wong/Projects/aind-behavior-video-transformation/src/aind_behavior_video_transformation/etl.py", line 273, in <module>
    job_response = job.run_job()
                   ^^^^^^^^^^^^^
  File "/Users/jonathan.wong/Projects/aind-behavior-video-transformation/src/aind_behavior_video_transformation/etl.py", line 150, in run_job
    self._run_compression(convert_video_args)
  File "/Users/jonathan.wong/Projects/aind-behavior-video-transformation/src/aind_behavior_video_transformation/etl.py", line 115, in _run_compression
    raise RuntimeError('One or more Ffmpeg jobs failed. See error logs.')
RuntimeError: One or more Ffmpeg jobs failed. See error logs.
```

- Log for 1 valid request and 1 invalid request:
```
2024-11-08 13:23:24,716 - INFO - FFmpeg job completed: /private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmphns1wvcc/camera2/clip.mp4
2024-11-08 13:23:24,730 - ERROR - FFmpeg conversion failed for /private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmpl9o2d0kg/camera1/clip.mp4
Command: ffmpeg -y -v warning -hide_banner invalid user input args -i /private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmpl9o2d0kg/camera1/clip.mp4 invalid user output args /private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmphns1wvcc/camera1/clip.mp4
Return code: 234
Error output:
[AVFormatContext @ 0x1341056c0] Unable to choose an output format for 'invalid'; use a standard extension for the filename or specify the format manually.
[out#0 @ 0x1341105d0] Error initializing the muxer for invalid: Invalid argument
Error opening output file invalid.
Error opening output files: Invalid argument
```

- Log for 2 invalid requests: 
```
2024-11-08 13:27:04,397 - ERROR - FFmpeg conversion failed for /private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmpoco639kw/camera1/clip.mp4
Command: ffmpeg -y -v warning -hide_banner invalid user input args -i /private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmpoco639kw/camera1/clip.mp4 invalid user output args /private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmpm5pxsmq7/camera1/clip.mp4
Return code: 234
Error output:
[AVFormatContext @ 0x122e2ded0] Unable to choose an output format for 'invalid'; use a standard extension for the filename or specify the format manually.
[out#0 @ 0x122e38d40] Error initializing the muxer for invalid: Invalid argument
Error opening output file invalid.
Error opening output files: Invalid argument

2024-11-08 13:27:04,397 - ERROR - FFmpeg conversion failed for /private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmpoco639kw/camera2/clip.mp4
Command: ffmpeg -y -v warning -hide_banner invalid user input args -i /private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmpoco639kw/camera2/clip.mp4 invalid user output args /private/var/folders/jp/tzg4yhb959s7g46rlqjjl2qh0000gp/T/tmpm5pxsmq7/camera2/clip.mp4
Return code: 234
Error output:
[AVFormatContext @ 0x14cf07720] Unable to choose an output format for 'invalid'; use a standard extension for the filename or specify the format manually.
[out#0 @ 0x14cf12660] Error initializing the muxer for invalid: Invalid argument
Error opening output file invalid.
Error opening output files: Invalid argument
```

**Comment by galenlynch** [comment]
I see, so if _any_ job fails then the relevant errors will be in the log?

This looks great! I will give it a once over one last time tomorrow.

**Review by galenlynch — APPROVED** [review]
Looks great!
